From c2c864a53ba7d0577267c4a69590bd9a540ced58 Mon Sep 17 00:00:00 2001
From: Huy Minh <buingoc67@gmail.com>
Date: Sat, 1 Jun 2024 09:04:02 +0000
Subject: [PATCH] Extra LSM count for KernelSU (v2)

Or else the kernel will go panic with "Too many LSMs registered"

v2: Updated due to new 6.12 changes

Signed-off-by: Huy Minh <buingoc67@gmail.com>
---
 include/linux/args.h      | 6 +++---
 include/linux/lsm_count.h | 9 ++++++++-
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/include/linux/args.h b/include/linux/args.h
index 2e8e65d975c7..2d3bd5aaba06 100644
--- a/include/linux/args.h
+++ b/include/linux/args.h
@@ -17,9 +17,9 @@
  * that as _n.
  */
 
-/* This counts to 15. Any more, it will return 16th argument. */
-#define __COUNT_ARGS(_0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _n, X...) _n
-#define COUNT_ARGS(X...) __COUNT_ARGS(, ##X, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+/* This counts to 16. Any more, it will return 17th argument. */
+#define __COUNT_ARGS(_0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _n, X...) _n
+#define COUNT_ARGS(X...) __COUNT_ARGS(, ##X, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
 
 /* Concatenate two parameters, but allow them to be expanded beforehand. */
 #define __CONCAT(a, b) a ## b
diff --git a/include/linux/lsm_count.h b/include/linux/lsm_count.h
index 16eb49761b25..dac433c837b4 100644
--- a/include/linux/lsm_count.h
+++ b/include/linux/lsm_count.h
@@ -102,6 +102,12 @@
 #define IPE_ENABLED
 #endif
 
+#if IS_ENABLED(CONFIG_KSU)
+#define KSU_ENABLED 1,
+#else
+#define KSU_ENABLED
+#endif
+
 /*
  *  There is a trailing comma that we need to be accounted for. This is done by
  *  using a skipped argument in __COUNT_LSMS
@@ -124,7 +130,8 @@
 		LANDLOCK_ENABLED	\
 		IMA_ENABLED		\
 		EVM_ENABLED		\
-		IPE_ENABLED)
+		IPE_ENABLED		\
+		KSU_ENABLED)
 
 #else
 
-- 
2.34.1